plugins/references.py

```diff
@@ -760,6 +781,70 @@ def run(self, edit, name=None):
             offset -= convert2ref(view, edit, _link_span, link_span[1], link_span[2])
 
 
+class MdeConvertBareLinkToMdLinkCommand(MdeTextCommand):
+    """Convert an inline link to reference."""
+
+    def is_visible(self):
+        """Return True if selection contains links"""
+        view = self.view
+        for sel in view.find_by_selector(barelink_scope_name):
+            if any(s.intersects(sel) for s in view.sel()):
+                return True
+        return False
+
+    def run(self, edit, name=None):
+        """Run command callback."""
+        # import queue
+        import threading
+        import time
+
+        thread_queue = []
+
+        url_titles = {}
+        url_redirects = {}
+
+        def getTitleFromUrlJob(link_href):
+            import urllib.request
+
+            resp = urllib.request.urlopen(link_href)
+            match = re.search(rb"<title>(.+?)</title>", resp.read())
+            if match:
+                url_titles[link_href] = re.sub(r"([\[\]])", r"\\\g<1>", match.group(1).decode())
+
+            real_url = resp.geturl()
+            if real_url != link_href:
+                print(link_href, "=/=", real_url, "Redirect?")
+                url_redirects[link_href] = real_url
+
+        view = self.view
+        valid_regions = find_by_selector_in_regions(view, view.sel(), barelink_scope_name)
+
+        for link_region in valid_regions:
+            link_href = view.substr(link_region)
+            thread_queue.append(threading.Thread(target=getTitleFromUrlJob, args=(link_href,)))
+
+        while True:
+            left = len([thread for thread in thread_queue if thread.is_alive()])
+            view.set_status("rawlinktomd", "Fetching " + str(left) + " pages")
+            if left == 0:
+                break
+            time.sleep(0.2)
```

Comment:
Converting multiple links can easily take several seconds and cause the GUI to become unresponsive. Especially on Windows it then causes the themed titlebar of ST to be replaced by the default one. It looks like ST has crashed.

Running those requests in threads doesn't seem to help much due to python's interpreter lock and/or the plugin host's main thread being blocked by `sleep` - not sure.

While I like the general idea, blocking the main UI thread isn't a good idea with regards to user experience.

Long running tasks need to be asynchronous. Either by implementing our own asyncio event loop or by scheduling tasks with something like https://github.com/jisaacks/GitGutter/blob/master/modules/tasks.py

Reply:
This definitely needs to be made synchronous for a release. What you're seeing here is some scaffolding for an asynchronous job queue, but as you've seen the UI is currently blocked, which isn't good. I wasn't sure how you wanted to handle that, since I wasn't yet familiar with how ST plugins handle async

Reply:
I haven't made up my mind about a solution, yet, as I am currently focused on syntax definition. 

The references module is quite large and could probably benefit from newer ST APIs such as `find_by_selector` etc. which could help to avoid several regexps, but require a reliable syntax. The module may also be merged with footnotes module as both share some functionalities or use different implementations for nearly the same things.

Reply:
Async seems particularly tricky here because of `Edit objects may not be used after the TextCommand's run method has returned`

Reply:
Final text manipulation needs to be performed by a dedicated TextCommand once the async task has finished, yes.